ta4j-core/src/main/java/org/ta4j/core/indicators/averages/JMAIndicator.java

```diff
@@ -0,0 +1,137 @@
+ *
+ * JMA, or Jurik Moving Average, is a type of moving average developed by Mark
+ * Jurik. It is known for its ability to respond to price changes more smoothly
+ * than traditional moving averages like SMA (Simple Moving Average) or EMA
+ * (Exponential Moving Average), while avoiding much of the lag associated with
+ * those averages.
+ *
+ */
+public class JMAIndicator extends CachedIndicator<Num> {
+
+    private final Indicator<Num> indicator; // Base indicator (e.g., Close Price)
+    private final int barCount; // Period for the JMA calculation
+    private final Num phase; // Phase adjustment (usually between -100 and +100)
+    private final Num power; // Smoothing power factor (default is 2)
+    private final Map<Integer, JmaData> jmaDataMap;
+    private final BarSeries barSeries;
+    private final NumFactory numFactory;
+    private final Num beta;
+    private final Num phaseRatio;
+    private final Num alpha;
+
+    /**
+     * Constructor.
+     *
+     * @param indicator The input indicator (e.g., Close Price).
+     * @param barCount  The number of periods for the JMA.
+     * @param phase     The phase adjustment (-100 to +100).
+     * @param power     The smoothing power factor (default is 2).
+     */
+    public JMAIndicator(Indicator<Num> indicator, int barCount, double phase, double power) {
+        super(indicator.getBarSeries());
+        this.barSeries = indicator.getBarSeries();
+        this.numFactory = barSeries.numFactory();
+        this.indicator = indicator;
+        this.barCount = barCount;
+        this.phase = numFactory.numOf(Math.min(Math.max(phase, -100), 100)); // Clamp phase between -100 and 100
+        this.power = numFactory.numOf(Math.max(1, power)); // Ensure power is at least 1
+        this.jmaDataMap = new HashMap<>();
+
+        // Compute smoothing factor based on phase
+        beta = numFactory.numOf(0.45)
+                .multipliedBy(numFactory.numOf(barCount - 1))
+                .dividedBy(numFactory.numOf(0.45).multipliedBy(numFactory.numOf(barCount - 1)).plus(numFactory.two()));
+
+        phaseRatio = this.phase.isLessThan(numFactory.numOf(-100)) ? numFactory.numOf(0.5)
+                : (this.phase.isGreaterThan(numFactory.numOf(100)) ? numFactory.numOf(2.5)
+                        : this.phase.dividedBy(numFactory.numOf(100)).plus(numFactory.numOf(1.5)));
+
+        alpha = beta.pow(this.power);
+
+        for (int i = 0; i < indicator.getBarSeries().getBarCount(); i++) {
+            calculate(i);
+        }
+    }
+
+    @Override
+    protected Num calculate(int index) {
+
+        NumFactory numFactory = indicator.getBarSeries().numFactory();
+        Num currentPrice = indicator.getValue(index);
+
+        if (index <= 0) {
+            jmaDataMap.put(index, new JmaData(currentPrice, numFactory.zero(), numFactory.zero(), currentPrice));
+            return currentPrice;
+        }
+
+        JmaData previousJMA = jmaDataMap.get(index - 1);
```

Comment:
This may work for backtesting, but causes memory leak within long running livetrading.

I can't see clean up after window passes further.

Reply:
I wouldn't classify this as a memory leak really.  It's a memory growth issue.   However, I did look at this more and attempted to extract E0, E1 and E2 into their own indicators and that didn't work either since E2 is dependent on JMA which would introduce a circular dependency.   I also looked at limited the jmaDataMap to a fixed size and that doesn't work either since any other reference to a previous JMA would fail. 

If you have any ideas on a different approach here let me know and I'll look into it.

The E2 calculation is the sticking point here since that references E0, E1, previous_E2 and previous_JMA.

e2 = (e0 + phaseRatio * e1 - JMA_prev) * (1 - alpha)^2 + alpha^2 * e2_prev